yt/utilities/lib/particle_kdtree_tools.pyx

```diff
@@ -84,48 +113,210 @@ def generate_smoothing_length(np.float64_t[:, ::1] input_positions,
 
             # Fill queue with particles in the node containing the particle
             # we're searching for
-            process_node_points(leafnode, pos, input_positions, queue, i)
+            process_node_points(leafnode, queue, input_positions, pos, i,
+                                &axes)
 
             # Traverse the rest of the kdtree to finish the neighbor list
-            find_knn(
-                c_tree.root, queue, input_positions, pos, leafnode.leafid, i)
+            find_knn(c_tree.root, queue, input_positions, pos, leafnode.leafid,
+                     i, &axes)
 
             smoothing_length[i] = sqrt(queue.heap_ptr[0])
 
     pbar.update(n_particles-1)
     pbar.finish()
     return np.asarray(smoothing_length)
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def query(PyKDTree kdtree, np.float64_t [:, ::1] tree_positions,
+          np.float64_t[:, ::1] input_positions, int num_neigh
+          ):
+    """This is a KD nearest neighbor search with a similar interface to the
+    scikdtree used for performance comparisons
+
+    Parameters
+    ----------
+
+    kdtree: A PyKDTree instance
+        A kdtree to do nearest neighbors searches with
+    tree_positions: arrays of floats with shape (n_particles, 3)
+        The positions of particles in kdtree non-sorted order. Currently assumed
+        to be 3D postions.
+    input_positions: array of floats (any, 3)
+        The positions to gather the nearest neighbors.
+    kdtree: A PyKDTree instance
+        A kdtree to do nearest neighbors searches with
+    num_neigh: The neighbor number to calculate the distance to
+
+    Returns
+    -------
+
+    dists: arrays of floats with shape (n_particles, num_neigh)
+        The the nearest neighbor distances
+    pids: arrays of ints with shape (n_particles, num_neigh)
+        The particle ids of the nearest neighbors
+    """
+
+    cdef KDTree* c_tree = kdtree._tree
+    cdef Node* leafnode
+    cdef np.float64_t* pos
+    cdef int i, skipidx = -1
+    cdef BoundedPriorityQueue queue = BoundedPriorityQueue(num_neigh, True)
+    cdef np.float64_t[:, :] dists
+    cdef np.int64_t[:, :] pids
+
+    cdef axes_range axes
+    set_axes_range(&axes, -1)
+
+    dists = np.zeros((input_positions.shape[0], num_neigh), dtype="float64")
+    pids = np.zeros((input_positions.shape[0], num_neigh), dtype="int64")
```

Comment:
Do these need to be signed?

Reply:
They don't *need* to be, but conventionally they can. We start with an invalidated queue and heap set to -1. If for some reason we didn't find 60 nearest neighbors, i.e, if filling chunkwise we could store intermediate negative values into both of these arrays. That convention could be changed though.

Reply:
Okay, just checking -- if you are using negative numbers then go ahead.